src/stdlib_experimental_stats.md

```diff
@@ -0,0 +1,43 @@
+# Descriptive statistics
+
+## Implemented
+
+ * `mean`
+
+## MEAN - mean of array elements
+
+### Description:
+
+Returns the mean of all the elements of *array*, or of the elements of *array* along dimension *dim*.
+
+### Syntax:
+
+RESULT = mean(*array*)
+
+RESULT = mean(*array*, *dim*)
+
+### Arguments:
+
+*array*: Shall be an array of type INTEGER, or REAL.
+
+*dim* (optional): Shall be a scalar of type INTEGER with a value in the range from 1 to n, where n is the rank of *array*.
```

Comment:
Ditto for formatting here as elsewhere in this file.

In addition: Should we call `dim` optional here? Technically it's not declared as optional, even though it may appear so to the caller. The important difference is that the user is not allowed to pass an absent optional dummy argument to `mean()` as `dim` because `dim` is not declared as optional. Modern Fortran Explained makes this distinction by saying that `sum` (and friends) can be invoked in one of two forms:

1. Without second argument `dim`;
2. With second argument `dim`.

Considering this is a spec, we should be strict in the language. What do you think?

Reply:
Another formatting nit-pick: Don't need colon (:) in subsection titles.

Reply:
> In addition: Should we call `dim` optional here? Technically it's not declared as optional, even though it may appear so to the caller. The important difference is that the user is not allowed to pass an absent optional dummy argument to `mean()` as `dim` because `dim` is not declared as optional. Modern Fortran Explained makes this distinction by saying that `sum` (and friends) can be invoked in one of two forms:
> 
> 1. Without second argument `dim`;
> 2. With second argument `dim`.
> 
> Considering this is a spec, we should be strict in the language. What do you think?

@milancurcic  Good point.

From [Standard Fortran draft](http://isotc.iso.org/livelink/livelink?func=ll&objId=19442438&objAction=Open&viewType=1):

* Page xviii:
> In references to the intrinsic functions ALL, ANY, FINDLOC, IALL, IANY, IPARITY, MAXLOC, MAXVAL, MINLOC, MINVAL, NORM2, PARITY, PRODUCT, SUM, and THIS_IMAGE, the actual argument for DIM can be a present optional dummy argument.

* Page 429:

> 16.9.185 SUM (ARRAY, DIM [, MASK]) or SUM (ARRAY [, MASK])
> 1 Description. Array reduced by addition.
> 2 Class. Transformational function.
> 3 Arguments.
> ARRAY shall be an array of numeric type.
> DIM shall be an integer scalar with a value in the range 1 ≤ DIM ≤ n, where n is the rank of ARRAY.
> MASK (optional) shall be of type logical and shall be conformable with ARRAY.

So clearly, `dim` is not considered as `optional` as it is for `mask`.

So, our spec should stated probably something like:

> `dim` : Shall be a scalar of type `integer` with a value in the range from 1 to n, where n is the rank of `array`.


This will be also more in agreement with the section "Syntax" where it is mentioned:

>`result = mean(array)`
>`result = mean(array, dim)`

instead of

>`result = mean(array [, dim])`

as traditionally used for optional argument.

Does it make sense?

Reply:
Great! The Description and Syntax section are clear (to me) that `mean` can be called with and without `dim`. So I think your suggestion is spot on, to just drop the `optional` from the Arguments section.